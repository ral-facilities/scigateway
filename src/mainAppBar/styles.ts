const appBarIconButtonStyle = {
  margin: 1,
  color: 'primary.contrastText',
};

const appBarMenuItemIconStyle = {
  marginLeft: '-12px',
  marginRight: 0,
  color: 'primary.contrastText',
};

export { appBarIconButtonStyle, appBarMenuItemIconStyle };
