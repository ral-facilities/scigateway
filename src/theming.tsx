import {
  createTheme,
  ThemeProvider,
  StyledEngineProvider,
  Theme,
} from '@mui/material/styles';
import 'react-redux-toastr/lib/css/react-redux-toastr.min.css';
import React from 'react';
import { StateType } from './state/state.types';
import { connect, useSelector } from 'react-redux';
import { checkboxClasses } from '@mui/material/Checkbox';
import {
  formHelperTextClasses,
  formLabelClasses,
  inputClasses,
  outlinedInputClasses,
  tabClasses,
} from '@mui/material';

/* UKRI colours */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const UKRI_COLOURS = {
  blue: '#2E2D62',
  bright: {
    orange: '#FF6900', // pure orange
    yellow: '#FBBB10', // yellow
    green: '#67C04D', // light green
    blue: '#1E5DF8', // blue
    purple: '#BE2BBB', // bright purple
    red: '#E94D36', // light red
  },
  deep: {
    orange: '#C13D33', // pure orange
    yellow: '#F08900', // vivid yellow
    green: '#3E863E', // green
    blue: '#003088', // blue
    purple: '#8A1A9B', // bright purple
    red: '#A91B2E', // red
  },
};

/* Colours that may be used across light/dark modes e.g. the main app bar */
const STATIC_COLOURS = {
  darkBlue: '#003088',
  orange: '#C34F00',
};

/* Main colours used for dark/light modes respectively */
interface ThemeColours {
  /* The type of mode this is */
  type: 'default' | 'contrast';

  /* Primary/secondary colours used for MUI */
  primary: string;
  secondary: string;

  /* Secondary text colour */
  textSecondary: string;

  /* Background colours for the page and papers */
  background: string;
  paper: string;

  /* Standard colours used in plugins (change to lighter/darker shades
    between light and dark modes) - these are meant to give good contrast
    for text on the chosen paper background colour */
  blue: string;
  orange: string;
  red: string;
  grey: string;

  /* Lighter colours */
  lightBlue: string; //Used for lighter coloured buttons
  lightOrange: string; //Used for notifcation icon

  /* Darker colours */
  darkGreen: string; //Used for cookie consent message
  darkOrange: string; //Used for help tour

  /* Contrast colours that need to change significantly between dark
     and light modes */
  contrastGrey: string; //Used on homepage

  /* Informational/Warning colours */
  information: string; //Used in open data label
  warning: string; //Used for selection alert banner

  /* Colours for <a> style links */
  link: {
    default: string;
    visited: string;
    active: string;
  };

  footerLink: {
    default: string;
    active: string;
  };

  chip: string; //Used for chip colours in cards and filters

  /* Colours used on the homepage */
  homePage: {
    heading: string;
    blueDescription: string;
    blueButton: string;
  };

  /* Colour for Tabs in search */
  tabsGrey: string;
}

declare module '@mui/material/styles' {
  interface Theme {
    colours: ThemeColours;
    drawerWidth: string;
    footerPaddingTop: string;
    footerPaddingBottom: string;
    footerHeight: string;
    mainAppBarHeight: string;
  }
  // allow configuration using `createTheme`
  interface ThemeOptions {
    colours: ThemeColours;
    drawerWidth: string;
    footerPaddingTop: string;
    footerPaddingBottom: string;
    footerHeight: string;
    mainAppBarHeight: string;
  }
}

const DARK_MODE_COLOURS: ThemeColours = {
  type: 'default',
  primary: UKRI_COLOURS.deep.blue,
  secondary: '#80ACFF',
  textSecondary: 'rgba(255, 255, 255, 0.7)',
  background: '#1B1B1B',
  paper: '#3A3A3A',
  blue: '#86B4FF',
  orange: '#F26300',
  red: '#FF7F73',
  grey: '#A4A4A4',
  lightBlue: UKRI_COLOURS.bright.blue,
  lightOrange: '#FF6900',
  darkGreen: '#3A7F3A',
  darkOrange: STATIC_COLOURS.orange,
  contrastGrey: '#E6E6E6',
  information: UKRI_COLOURS.deep.blue,
  warning: '#FFA500',
  link: {
    default: '#86B4FF',
    visited: '#BE2BBB',
    active: '#E94D36',
  },
  footerLink: {
    default: '#FFFFFF',
    active: '#E58885',
  },
  chip: '#595959',
  homePage: {
    heading: '#FFFFFF',
    blueDescription: '#FFFFFF',
    blueButton: UKRI_COLOURS.bright.blue,
  },
  tabsGrey: '#3A3A3A',
};

const DARK_MODE_HIGH_CONTRAST_COLOURS: ThemeColours = {
  type: 'contrast',
  primary: '#86B4FF',
  secondary: '#80ACFF',
  textSecondary: 'rgba(255, 255, 255, 0.7)',
  background: '#000000',
  paper: '#1A1A1A',
  blue: '#B4CCFA',
  orange: '#FFC14D',
  red: '#FFA198',
  grey: '#A4A4A4',
  lightBlue: UKRI_COLOURS.deep.blue,
  lightOrange: '#FFC14D',
  darkGreen: '#3A7F3A',
  darkOrange: STATIC_COLOURS.orange,
  contrastGrey: '#FFFFFF',
  information: UKRI_COLOURS.deep.blue,
  warning: '#FFC14D',
  link: {
    default: '#B4CCFA',
    visited: '#BE2BBB',
    active: '#E94D36',
  },
  footerLink: {
    default: '#000000',
    active: '#851D0F',
  },
  chip: '#3A3A3A',
  homePage: {
    heading: '#FFFFFF',
    blueDescription: '#FFFFFF',
    blueButton: UKRI_COLOURS.bright.blue,
  },
  tabsGrey: '#1A1A1A',
};

const LIGHT_MODE_COLOURS: ThemeColours = {
  type: 'default',
  primary: UKRI_COLOURS.deep.blue,
  secondary: UKRI_COLOURS.deep.blue,
  textSecondary: 'rgba(0, 0, 0, 0.54)',
  background: '#FAFAFA',
  paper: '#FFFFFF',
  blue: UKRI_COLOURS.deep.blue,
  orange: '#C34F00',
  red: '#AC1600',
  grey: '#727272',
  lightBlue: UKRI_COLOURS.bright.blue,
  lightOrange: '#FF6900',
  darkGreen: '#3A7F3A',
  darkOrange: STATIC_COLOURS.orange,
  contrastGrey: '#717171',
  information: UKRI_COLOURS.deep.blue,
  warning: '#FFA500',
  link: {
    default: '#1E5DF8',
    visited: '#BE2BBB',
    active: '#E94D36',
  },
  footerLink: {
    default: '#FFFFFF',
    active: '#E58885',
  },
  chip: '#E0E0E0',
  homePage: {
    heading: '#333333',
    blueDescription: '#FFFFFF',
    blueButton: UKRI_COLOURS.bright.blue,
  },
  tabsGrey: '#EEEEEE',
};

const LIGHT_MODE_HIGH_CONTRAST_COLOURS: ThemeColours = {
  type: 'contrast',
  primary: UKRI_COLOURS.deep.blue,
  secondary: UKRI_COLOURS.deep.blue,
  textSecondary: '#000000',
  background: '#FAFAFA',
  paper: '#FFFFFF',
  blue: '#002466',
  orange: '#C34F00',
  red: '#801100',
  grey: '#727272',
  lightBlue: UKRI_COLOURS.bright.blue,
  lightOrange: '#FF6900',
  darkGreen: '#3A7F3A',
  darkOrange: STATIC_COLOURS.orange,
  contrastGrey: '#000000',
  information: UKRI_COLOURS.deep.blue,
  warning: '#FFA500',
  link: {
    default: '#052d94',
    visited: '#BE2BBB',
    active: '#E94D36',
  },
  footerLink: {
    default: '#FFFFFF',
    active: '#E58885',
  },
  chip: '#E0E0E0',
  homePage: {
    heading: '#000000',
    blueDescription: '#FFFFFF',
    blueButton: UKRI_COLOURS.bright.blue,
  },
  tabsGrey: '#EEEEEE',
};

export const buildTheme = (
  darkModePreference: boolean,
  highContrastModePreference?: boolean,
  primaryColour?: string
): Theme => {
  const colours = darkModePreference
    ? highContrastModePreference
      ? DARK_MODE_HIGH_CONTRAST_COLOURS
      : DARK_MODE_COLOURS
    : highContrastModePreference
      ? LIGHT_MODE_HIGH_CONTRAST_COLOURS
      : LIGHT_MODE_COLOURS;

  if (primaryColour && !(darkModePreference && highContrastModePreference)) {
    colours.primary = primaryColour;
  }

  const componentOverrides = {
    MuiPaper: {
      styleOverrides: { root: { backgroundImage: 'unset' } },
    },
    MuiLink: {
      defaultProps: {
        underline: 'hover' as 'always' | 'none' | 'hover' | undefined,
      },
      styleOverrides: {
        root: {
          color: colours.blue,
        },
      },
    },
    MuiTabs: {
      styleOverrides: {
        indicator: {
          color: STATIC_COLOURS.darkBlue,
          textDecoration: 'underline',
        },
      },
    },
    MuiTab: {
      styleOverrides: {
        textColorSecondary: {
          [`&.${tabClasses.selected}`]: {
            color: darkModePreference ? '#FFFFFF' : colours.blue,
          },
        },
      },
    },
    MuiCheckbox: {
      styleOverrides: {
        colorPrimary: {
          [`&.${checkboxClasses.checked}`]: {
            color: colours.blue,
          },
          [`&.${checkboxClasses.indeterminate}`]: {
            color: colours.grey,
          },
        },
      },
    },
    MuiFormLabel: {
      styleOverrides: {
        root: {
          [`&.${formLabelClasses.error}`]: {
            color: colours.red,
          },
          [`&.${formLabelClasses.focused}`]: {
            color: colours.blue,
          },
        },
        asterisk: {
          [`&.${formLabelClasses.error}`]: {
            color: colours.red,
          },
        },
      },
    },
    MuiBadge: {
      styleOverrides: {
        colorPrimary: {
          backgroundColor: STATIC_COLOURS.orange,
          color: 'white',
        },
      },
    },
    MuiInput: {
      styleOverrides: {
        underline: {
          [`&.${inputClasses.error}`]: {
            borderBottomColor: colours.red,
          },
          [`&.${inputClasses.error}:after`]: {
            borderBottomColor: colours.blue,
          },
        },
      },
    },
    MuiOutlinedInput: {
      styleOverrides: {
        root: {
          [`&.${outlinedInputClasses.error} .${outlinedInputClasses.notchedOutline}`]:
            {
              borderColor: colours.red,
            },
        },
        //Override opacity of placeholder text
        input: {
          '&::placeholder': {
            colour: colours.textSecondary,
            opacity:
              !darkModePreference && highContrastModePreference ? 1.0 : 0.7,
          },
        },
      },
    },
    MuiFormHelperText: {
      styleOverrides: {
        root: {
          [`&.${formHelperTextClasses.error}`]: {
            color: colours.red,
          },
        },
      },
    },
    MuiChip: {
      styleOverrides: {
        root: {
          backgroundColor: colours.chip,
        },
      },
    },
    MuiPickersToolbar: {
      styleOverrides: {
        toolbar: {
          backgroundColor: colours.primary,
        },
      },
    },
    MuiPickersCalendarHeader: {
      styleOverrides: {
        dayLabel: {
          color: colours.grey,
        },
      },
    },
    MuiPickersDay: {
      styleOverrides: {
        current: {
          color: colours.blue,
        },
        dayDisabled: {
          color: colours.grey,
        },
      },
    },
    MuiPickersYear: {
      styleOverrides: {
        root: {
          '&:active': {
            color: colours.blue,
          },
          '&:focus': {
            color: colours.blue,
          },
        },
        yearSelected: {
          color: colours.blue,
        },
        yearDisabled: {
          color: colours.grey,
        },
      },
    },
    MuiPickersMonth: {
      styleOverrides: {
        root: {
          '&:active': {
            color: colours.blue,
          },
          '&:focus': {
            color: colours.blue,
          },
        },
        monthSelected: {
          color: colours.blue,
        },
        monthDisabled: {
          color: colours.grey,
        },
      },
    },
  };

  return createTheme({
    palette: {
      // Light/dark mode
      mode: darkModePreference ? 'dark' : 'light',
      primary: {
        main: colours.primary,
      },
      secondary: {
        main: colours.secondary,
      },
      text: {
        secondary: colours.textSecondary,
      },
      background: {
        default: colours.background,
        paper: colours.paper,
      },
    },
    typography: {
      button: {
        textTransform: 'none',
      },
    },
    components: componentOverrides,
    colours: colours,
    drawerWidth: '220px',
    footerPaddingTop: '8px',
    footerPaddingBottom: '8px',
    footerHeight: '20px',
    mainAppBarHeight: '64px',
  });
};

function mapThemeProviderStateToProps(state: StateType): {
  prefersDarkMode: boolean;
} {
  return {
    prefersDarkMode: state.scigateway.darkMode,
  };
}

const SciGatewayThemeProvider = (props: {
  children: React.ReactNode;
}): React.ReactElement<{
  children: React.ReactNode;
}> => {
  const darkModePreference = useSelector(
    (state: StateType) => state.scigateway.darkMode
  );
  const highContrastModePreference = useSelector(
    (state: StateType) => state.scigateway.highContrastMode
  );
  const primaryColour = useSelector(
    (state: StateType) => state.scigateway.primaryColour
  );
  return (
    <StyledEngineProvider injectFirst>
      <ThemeProvider
        theme={buildTheme(
          darkModePreference,
          highContrastModePreference,
          primaryColour
        )}
      >
        {props.children}
      </ThemeProvider>
    </StyledEngineProvider>
  );
};

export const ConnectedThemeProvider = connect(mapThemeProviderStateToProps)(
  SciGatewayThemeProvider
);
