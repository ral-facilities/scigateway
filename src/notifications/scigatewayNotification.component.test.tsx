import React from 'react';
import Notification from './scigatewayNotification.component';
import { StyledEngineProvider, ThemeProvider } from '@mui/material/styles';
import { buildTheme } from '../theming';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

function createScigatewayNotification(
  severity: string,
  message: string
): React.ReactElement {
  return (
    <Notification
      message={message}
      severity={severity}
      index={0}
      dismissNotification={jest.fn()}
    />
  );
}

describe('Scigateway Notification component', () => {
  function Wrapper({
    children,
  }: {
    children: React.ReactElement;
  }): JSX.Element {
    return (
      <StyledEngineProvider injectFirst>
        <ThemeProvider theme={buildTheme(false)}>{children}</ThemeProvider>
      </StyledEngineProvider>
    );
  }

  it('Scigateway Notification success message renders correctly', () => {
    const { asFragment } = render(
      createScigatewayNotification('success', 'success message'),
      { wrapper: Wrapper }
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('Scigateway Notification warning message renders correctly', () => {
    const { asFragment } = render(
      createScigatewayNotification('warning', 'warning message'),
      { wrapper: Wrapper }
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('Scigateway Notification error message renders correctly', () => {
    const { asFragment } = render(
      createScigatewayNotification('error', 'error message'),
      { wrapper: Wrapper }
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('Scigateway Notification information message renders correctly', () => {
    const { asFragment } = render(
      createScigatewayNotification('information', 'information message'),
      { wrapper: Wrapper }
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('an action is fired when Scigateway Notification button is clicked', async () => {
    const mockDismissFn = jest.fn();
    const user = userEvent.setup();

    render(
      <Notification
        message="warning message"
        severity="warning"
        index={0}
        dismissNotification={mockDismissFn}
      />,
      { wrapper: Wrapper }
    );

    await user.click(
      screen.getByRole('button', { name: 'Dismiss notification' })
    );

    expect(mockDismissFn.mock.calls.length).toEqual(1);
  });
});
