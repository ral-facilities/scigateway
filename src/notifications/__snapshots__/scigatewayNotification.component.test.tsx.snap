// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scigateway Notification component Scigateway Notification error message renders correctly 1`] = `
<DocumentFragment>
  <div
    style="display: flex; align-items: center;"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1867c3f-MuiSvgIcon-root"
      data-testid="HighlightOffIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M14.59 8 12 10.59 9.41 8 8 9.41 10.59 12 8 14.59 9.41 16 12 13.41 14.59 16 16 14.59 13.41 12 16 9.41 14.59 8zM12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
      />
    </svg>
    <p
      class="MuiTypography-root MuiTypography-body2 css-rsneeo-MuiTypography-root"
    >
      error message
    </p>
    <button
      aria-label="Dismiss notification"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-1qpmkg3-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1nrt62k-MuiSvgIcon-root"
        data-testid="ClearIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`Scigateway Notification component Scigateway Notification information message renders correctly 1`] = `
<DocumentFragment>
  <div
    style="display: flex; align-items: center;"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-kjenk6-MuiSvgIcon-root"
      data-testid="InfoIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
      />
    </svg>
    <p
      class="MuiTypography-root MuiTypography-body2 css-rsneeo-MuiTypography-root"
    >
      information message
    </p>
    <button
      aria-label="Dismiss notification"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-1qpmkg3-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1nrt62k-MuiSvgIcon-root"
        data-testid="ClearIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`Scigateway Notification component Scigateway Notification success message renders correctly 1`] = `
<DocumentFragment>
  <div
    style="display: flex; align-items: center;"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-18vlyb1-MuiSvgIcon-root"
      data-testid="CheckCircleIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
      />
    </svg>
    <p
      class="MuiTypography-root MuiTypography-body2 css-rsneeo-MuiTypography-root"
    >
      success message
    </p>
    <button
      aria-label="Dismiss notification"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-1qpmkg3-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1nrt62k-MuiSvgIcon-root"
        data-testid="ClearIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`Scigateway Notification component Scigateway Notification warning message renders correctly 1`] = `
<DocumentFragment>
  <div
    style="display: flex; align-items: center;"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1o88wul-MuiSvgIcon-root"
      data-testid="ErrorOutlineIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
      />
    </svg>
    <p
      class="MuiTypography-root MuiTypography-body2 css-rsneeo-MuiTypography-root"
    >
      warning message
    </p>
    <button
      aria-label="Dismiss notification"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-1qpmkg3-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1nrt62k-MuiSvgIcon-root"
        data-testid="ClearIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
